### DISCUSSION TITLE
Is it possible to use heatwave with extensions built off the develop branch?

### DISCUSSION THREAD

**Discussion by chrisbednarski** [wixtoolset/issues#7621] [Questions] [answered]
I'm using HeatWave 4.0.1 on VS 2022 (17.6.5). Thanks for this great tool.

I've forked the develop branch of wix and made some local changes to the firewall extension.
I'm trying to test those changes with my project, but I'm having trouble consuming the custom built nuget package (ie 5.0.0 preview ) from Visual Studio.

Is it possible to use these with HeatWave, or do all package versions have to match the HeatWave version?

After updating the `WixToolset.Firewall.wixext` PackageReference to the updated version in my project, Visual Studio is reporting the following error when building it:

wix.exe : error WIX0144: The extension 'C:\Users\[user.name]\.nuget\packages\wixtoolset.util.wixext\5.0.0-preview.1-build.30\wixext4\WixToolset.Util.wixext.dll' could not be loaded because of the following reason: System.IO.FileNotFoundException: Could not load file or assembly 'WixToolset.Extensibility, Version=5.0.0.0, Culture=neutral, PublicKeyToken=a7d136314861246c' or one of its dependencies. The system cannot find the file specified.

Fusion logs are showing that wix.exe 4.0.0 is failing to find the new version of `WixToolset.Extensibility` and `WixToolset.Data` 
![image](https://github.com/wixtoolset/issues/assets/1804713/e78b2b01-9413-45ad-bb38-504082701a46)

```
*** Assembly Binder Log Entry  (18/07/2023 @ 7:35:10 PM) ***

The operation failed.
Bind result: hr = 0x80070002. The system cannot find the file specified.

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework64\v4.0.30319\clr.dll
Running under executable  C:\Users\user.name\.nuget\packages\wixtoolset.sdk\4.0.0\tools\net472\x64\wix.exe
--- A detailed error log follows. 

=== Pre-bind state information ===
LOG: DisplayName = WixToolset.Extensibility, Version=5.0.0.0, Culture=neutral, PublicKeyToken=a7d136314861246c
 (Fully-specified)
LOG: Appbase = file:///C:/Users/user.name/.nuget/packages/wixtoolset.sdk/4.0.0/tools/net472/x64/
LOG: Initial PrivatePath = NULL
LOG: Dynamic Base = NULL
LOG: Cache Base = NULL
LOG: AppName = wix.exe
Calling assembly : WixToolset.Util.wixext, Version=5.0.0.0, Culture=neutral, PublicKeyToken=a7d136314861246c.
===
LOG: This bind starts in LoadFrom load context.
WRN: Native image will not be probed in LoadFrom context. Native image will only be probed in default load context, like with Assembly.Load().
LOG: Using application configuration file: C:\Users\user.name\.nuget\packages\wixtoolset.sdk\4.0.0\tools\net472\x64\wix.exe.Config
LOG: Using host configuration file: 
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework64\v4.0.30319\config\machine.config.
LOG: Post-policy reference: WixToolset.Extensibility, Version=5.0.0.0, Culture=neutral, PublicKeyToken=a7d136314861246c
LOG: GAC Lookup was unsuccessful.
LOG: Attempting download of new URL file:///C:/Users/user.name/.nuget/packages/wixtoolset.sdk/4.0.0/tools/net472/x64/WixToolset.Extensibility.DLL.
LOG: Assembly download was successful. Attempting setup of file: C:\Users\user.name\.nuget\packages\wixtoolset.sdk\4.0.0\tools\net472\x64\WixToolset.Extensibility.dll
LOG: Entering run-from-source setup phase.
LOG: Assembly Name is: WixToolset.Extensibility, Version=4.0.0.0, Culture=neutral, PublicKeyToken=a7d136314861246c
WRN: Comparing the assembly name resulted in the mismatch: Major Version
ERR: The assembly reference did not match the assembly definition found.
ERR: Run-from-source setup phase failed with hr = 0x80131040.
```
Is there a better way of testing and iterating my local changes?

**Comment by chrisbednarski** [comment]
Workaround:
I've changed the version on my develop branch of wix from 5.0.0 to 4.0.1, to match the version of HeatWave.
This is working, but with a lot of friction. eg. requires clearing the nuget cache with every custom 4.0.1 build I make and want to test.

Is there a better way of doing this?

**Comment by barnson** [comment]
What happens if you try using WixToolset.Sdk v5?

**Comment by chrisbednarski** [comment]
Thanks for your quick reply @barnson 

I'm not 100% sure what you mean. Do you mean [the command line](https://github.com/wixtoolset/wix/blob/develop/README.md?plain=1#L49) ? The command line works, ie shows the help. I haven't tried building my project via the command line, if that's what you're asking

How do I use WixToolset.Sdk v5 in VS 2022 ?

**Comment by chrisbednarski** [comment]
On a side note, the `devbuild.cmd` works well. I've also been calling `/src/ext/firewall/firewall.cmd` directly, to build just my changes. And this works well too.

However, I'm not able to build the `Firewall.winext` solution directly in Visual Studio.
- firewall project says it's incompatible

![image](https://github.com/wixtoolset/issues/assets/1804713/43e692a5-dfb4-4711-b109-c810e5dd35f7)

```
Rebuild started...
Restored C:\git\wix\src\ext\Firewall\wixext\WixToolset.Firewall.wixext.csproj (in 27 ms).
Restored C:\git\wix\src\ext\Firewall\test\WixToolsetTest.Firewall\WixToolsetTest.Firewall.csproj (in 156 ms).
1>------ Rebuild All started: Project: fwca, Configuration: Debug Win32 ------
2>------ Rebuild All started: Project: WixToolset.Firewall.wixext, Configuration: Debug Any CPU ------
2>CSC : error CS1566: Error reading resource 'WixToolset.Firewall.firewall.wixlib' -- 'Could not find file 'C:\git\wix\build\Firewall.wixext\Debug\firewall.wixlib'.'
2>Done building project "WixToolset.Firewall.wixext.csproj" -- FAILED.
3>------ Rebuild All started: Project: WixToolsetTest.Firewall, Configuration: Debug Any CPU ------
1>dllmain.cpp
3>CSC : error CS0006: Metadata file 'C:\git\wix\build\Firewall.wixext\Debug\netstandard2.0\WixToolset.Firewall.wixext.dll' could not be found
3>Done building project "WixToolsetTest.Firewall.csproj" -- FAILED.
1>firewall.cpp
1>   Creating library C:\git\wix\build\Firewall.wixext\Debug\x86\fwca.lib and object C:\git\wix\build\Firewall.wixext\Debug\x86\fwca.exp
1>fwca.vcxproj -> C:\git\wix\build\Firewall.wixext\Debug\x86\fwca.dll
========== Rebuild All: 1 succeeded, 2 failed, 0 skipped ==========
========== Rebuild started at 6:53 AM and took 02.971 seconds ==========
```

**Reply by robmen** [reply, marked as answer]
Loading the dev build of the WiX SDK within the WiX environment introduces its own unique challenges. We've not prioritized supporting that scenario in HeatWave.

Therefore, if you `Rebuild` you are deleting the .wixlib that won't get built, and that causes the rest to fail. So, try using `Build` instead if you want to use VS.

We haven't done a lot of individual .wixext development so the process (from WiX v4) is more optimized for changes across all .wixexts.

**Reply by chrisbednarski** [reply]
Thanks for this tip @robmen . Worked like a charm for building the extensions individually

TLDR
- HeatWave 4.0.1 does not support running dev builds
- build wix using `devbuild.cmd` first
- then, open individual wix solutions in VS and use incremental builds (ie. do not rebuild all)

**Comment by barnson** [comment]
I mostly build from the command line but you can specify a version in the `Sdk` attribute of a .wixproj to use, for example, `5.0.0-preview.1-build.29` from the GitHub feed or a local build.
